ui/app/scripts/stream/services/editor-service.js

```diff
@@ -0,0 +1,818 @@
+                });
+            }
+            if (closestData) {
+                return closestData;
+            }
+
+            // Check if drop on a link is allowed
+            var sourceType = source.attr('metadata/name');
+            var sourceGroup = source.attr('metadata/group');
+            if (targetUnderMouse instanceof joint.dia.Link && !(sourceType === 'destination' || sourceGroup === 'sink' || sourceGroup === 'source') && graph.getConnectedLinks(source).length === 0) {
+                return {
+                    context: context,
+                    source: {
+                        cell: source
+                    },
+                    target: {
+                        cell: targetUnderMouse
+                    }
+                };
+            }
+
+            return {
+                context: context,
+                source: {
+                    cell: source
+                },
+            };
+        }
+
+        function validateNode(flo, element) {
+            var errors = [];
+            var graph = flo.getGraph();
+            var group = element.attr('metadata/group');
+            var type = element.attr('metadata/name');
+            var incoming = [], outgoing = [], tap = [], invalidIncoming = [], invalidOutgoing = [];
+            var port;
+
+            graph.getConnectedLinks(element).forEach(function(link) {
+               if (link.get('source').id === element.id) {
+                   port = link.get('source').port;
+                   if (port === 'output') {
+                        outgoing.push(link);
+                   } else if (port === 'tap') {
+                       tap.push(link);
+                   } else {
+                       invalidOutgoing.push(link);
+                   }
+               } else if (link.get('target').id === element.id) {
+                   port = link.get('target').port;
+                   if (port === 'input') {
+                       incoming.push(link);
+                   } else {
+                       invalidIncoming.push(link);
+                   }
+               }
+            });
+
+            if (invalidIncoming.length > 0) {
+                errors.push({
+                    message: 'Invalid incoming link' + (invalidIncoming.length > 1 ? 's' : ''),
+                    range: element.attr('range')
+                });
+            }
+
+            if (invalidOutgoing.length > 0) {
+                errors.push({
+                    message: 'Invalid outgoing link' + (invalidOutgoing.length > 1 ? 's' : ''),
+                    range: element.attr('range')
+                });
+            }
+
+            if (group === 'source') {
+                if (incoming.length !== 0) {
+                    errors.push({
+                        message: 'Sources must appear at the start of a stream',
+                        range: element.attr('range')
+                    });
+                }
+                if (outgoing.length !== 1) {
```

Comment:
I think some of these code blocks can be moved to a separate function so they can be reused (especially for input/output)